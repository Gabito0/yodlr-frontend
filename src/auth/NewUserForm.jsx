import React, { useState } from "react";
import AppAlert from "../common/AppAlert";
import userApi from "../api/api";
import LoadingSpinner from "../common/LoadingSpinner";

/** New user creation form for admins. */
const NewUserForm = () => {
  const [formData, setFormData] = useState({
    email: "",
    firstName: "",
    lastName: "",
    password: "",
    isAdmin: false,
    state: "pending",
  });
  const [formErrors, setFormErrors] = useState([]);
  const [saveConfirmed, setSaveConfirmed] = useState(false);
  const [loading, setLoading] = useState(false); // State for loading

  /** Handle form submission */
  async function handleSubmit(evt) {
    evt.preventDefault();
    setLoading(true); // Start loading
    try {
      await userApi.createUser(formData);
      setFormData({
        email: "",
        firstName: "",
        lastName: "",
        password: "",
        isAdmin: false,
        state: "pending",
      });
      setFormErrors([]);
      setSaveConfirmed(true);
    } catch (errors) {
      setFormErrors(errors);
      setSaveConfirmed(false);
    } finally {
      setLoading(false); // Stop loading
    }
  }

  /** Handle form field change */
  function handleChange(evt) {
    const { name, value, type, checked } = evt.target;
    setFormData((f) => ({
      ...f,
      [name]: type === "checkbox" ? checked : value,
    }));
    setFormErrors([]);
  }

  return (
    <div className="container mt-5">
      <div className="row justify-content-center">
        <div className="col-md-8">
          <div className="card">
            <div className="card-header">
              <h3>Create New User</h3>
            </div>
            <div className="card-body">
              {loading ? (
                <LoadingSpinner />
              ) : (
                <form onSubmit={handleSubmit}>
                  <div className="form-group">
                    <label htmlFor="email">Email</label>
                    <input
                      id="email"
                      type="email"
                      name="email"
                      className="form-control"
                      value={formData.email}
                      onChange={handleChange}
                      autoComplete="email"
                      required
                    />
                  </div>
                  <div className="form-group">
                    <label htmlFor="firstName">First Name</label>
                    <input
                      id="firstName"
                      name="firstName"
                      className="form-control"
                      value={formData.firstName}
                      onChange={handleChange}
                      autoComplete="given-name"
                      required
                    />
                  </div>
                  <div className="form-group">
                    <label htmlFor="lastName">Last Name</label>
                    <input
                      id="lastName"
                      name="lastName"
                      className="form-control"
                      value={formData.lastName}
                      onChange={handleChange}
                      autoComplete="family-name"
                      required
                    />
                  </div>
                  <div className="form-group">
                    <label htmlFor="password">Password</label>
                    <input
                      id="password"
                      type="password"
                      name="password"
                      className="form-control"
                      value={formData.password}
                      onChange={handleChange}
                      autoComplete="new-password"
                      required
                    />
                  </div>
                  <div className="form-group form-check">
                    <input
                      id="isAdmin"
                      type="checkbox"
                      name="isAdmin"
                      className="form-check-input"
                      checked={formData.isAdmin}
                      onChange={handleChange}
                    />
                    <label htmlFor="isAdmin" className="form-check-label">
                      Admin Privileges
                    </label>
                  </div>

                  {formErrors.length ? (
                    <AppAlert type="danger" messages={formErrors} />
                  ) : null}

                  {saveConfirmed ? (
                    <AppAlert
                      type="success"
                      messages={["User created successfully."]}
                    />
                  ) : null}

                  <button className="btn btn-primary float-right" type="submit">
                    Create User
                  </button>
                </form>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default NewUserForm;
