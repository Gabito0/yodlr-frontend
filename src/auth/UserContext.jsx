import React from "react";

/**
 * UserContext provides a context for user information in the application.
 *
 * This context allows different parts of the app to access and update the current
 * user's information, such as their ID, email, and role (e.g., admin status).
 * It facilitates state management for user data across various components.
 *
 * Components can consume this context to access user data or to make updates
 * when user information changes, such as during login, logout, or profile updates.
 */

const UserContext = React.createContext();

export default UserContext;
