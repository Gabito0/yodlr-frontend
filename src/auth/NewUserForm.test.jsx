import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import NewUserForm from "./NewUserForm";
import UserContext from "./UserContext";
import userApi from "../api/api";
import { vitest, expect, it } from "vitest";

// Mock the userApi.createUser method
vi.mock("../api/api");

it("matches snapshot", function () {
  const mockUser = { currentUser: { isAdmin: true } };

  const { asFragment } = render(
    <MemoryRouter>
      <UserContext.Provider value={mockUser}>
        <NewUserForm />
      </UserContext.Provider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

it("submits form with valid data", async function () {
  const mockUser = { currentUser: { isAdmin: true } };
  userApi.createUser.mockResolvedValueOnce({});

  const { getByLabelText, getByText } = render(
    <MemoryRouter>
      <UserContext.Provider value={mockUser}>
        <NewUserForm />
      </UserContext.Provider>
    </MemoryRouter>
  );

  fireEvent.change(getByLabelText("Email"), {
    target: { value: "newuser@example.com" },
  });
  fireEvent.change(getByLabelText("First Name"), {
    target: { value: "New" },
  });
  fireEvent.change(getByLabelText("Last Name"), {
    target: { value: "User" },
  });
  fireEvent.change(getByLabelText("Password"), {
    target: { value: "password123" },
  });
  fireEvent.click(getByText("Admin Privileges"));

  fireEvent.click(getByText("Create User"));

  await waitFor(() => {
    expect(userApi.createUser).toHaveBeenCalledWith({
      email: "newuser@example.com",
      firstName: "New",
      lastName: "User",
      password: "password123",
      isAdmin: true,
      state: "pending",
    });
  });
});
