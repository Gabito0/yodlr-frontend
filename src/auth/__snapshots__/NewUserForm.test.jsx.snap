// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mt-5"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-md-8"
      >
        <div
          class="card"
        >
          <div
            class="card-header"
          >
            <h3>
              Create New User
            </h3>
          </div>
          <div
            class="card-body"
          >
            <form>
              <div
                class="form-group"
              >
                <label
                  for="email"
                >
                  Email
                </label>
                <input
                  autocomplete="email"
                  class="form-control"
                  id="email"
                  name="email"
                  required=""
                  type="email"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  for="firstName"
                >
                  First Name
                </label>
                <input
                  autocomplete="given-name"
                  class="form-control"
                  id="firstName"
                  name="firstName"
                  required=""
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  for="lastName"
                >
                  Last Name
                </label>
                <input
                  autocomplete="family-name"
                  class="form-control"
                  id="lastName"
                  name="lastName"
                  required=""
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  for="password"
                >
                  Password
                </label>
                <input
                  autocomplete="new-password"
                  class="form-control"
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
              </div>
              <div
                class="form-group form-check"
              >
                <input
                  class="form-check-input"
                  id="isAdmin"
                  name="isAdmin"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="isAdmin"
                >
                  Admin Privileges
                </label>
              </div>
              <button
                class="btn btn-primary float-right"
                type="submit"
              >
                Create User
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
